### ISSUE TITLE
Frequently hitting "error occurred while evaluating entity attributes: recursion limit reached" on AWS Lambda

### ISSUE THREAD

**Issue by tomtau** [cedar-policy/cedar#322]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/cedar-policy/cedar/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/cedar-policy/cedar/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### Bug Category

Policy Evaluation, Other

### Describe the bug

I ran Cedar policy (version 2.4.0) Authorizer for a small fixed policy document in a function with an Axum server on AWS Lambda (x86_64) and it frequently returned a response to `is_authorized` whose Diagnostics contained the following reason:
```
"error occurred while evaluating entity attributes: recursion limit reached" 
```

I could not reproduce it in local tests; it only happened in the AWS Lambda environment.

Likely due to this change https://github.com/cedar-policy/cedar/pull/8 by @aaronjeline @shaobo-he-aws @john-h-kastner-aws

### Expected behavior

Evaluate a small policy without hitting the recursion limit.

### Reproduction steps

1. run Axum server in AWS Lambda: https://github.com/awslabs/aws-lambda-rust-runtime/blob/main/examples/http-axum/ and add to one of its routes code that evaluates a small Cedar policy and log its response's diagnostics
2. invoke the lambda e.g. via API Gateway and check the logs

### Code Snippet

policy doc:
```
permit(principal, action == Action::"approve_or_reject", resource == AllDeploys::"");
permit(principal, action == Action::"list", resource == AllSystems::"");
permit(principal, action == Action::"new", resource == AllDeploys::"");
permit(principal, action == Action::"read", resource == AllDeploys::"")
```

example cedar policy evaluation:
```rust
    let session = r#"Session::"123""#.parse().unwrap(); // the session isn't 123, but a prefixed string with uuid, but that doesn't matter for the example, I guess
    let action = r#"Action::"list""#.parse().unwrap();
    let resource = r#"AllSystems::"""#.parse().unwrap();

    let entities = Entities::empty();

    let request = Request::new(Some(session), Some(action), Some(resource), Context::empty());

    let authorizer = Authorizer::new();
    let answer = authorizer.is_authorized(&request, &policy, &entities);
```

### Log output

"error occurred while evaluating entity attributes: recursion limit reached" 

### Operating  System

AWS Lambda (Amazon Linux 2 custom runtime / cargo-lambda)

### Additional information and screenshots

I was wondering whether there is any workaround and if there's something wrong with the way we wrote our Cedar policy: it's fairly small, so I wouldn't expect it to hit this recursion limit and I'm thinking what could be causing the issue... is it the empty strings in resources?

**Comment by aaronjeline** [comment]
I agree this seems far too small to be hitting the limit. How much ram does the lambda runner have? Are you compiling in release mode?

**Comment by tomtau** [comment]
> I agree this seems far too small to be hitting the limit. How much ram does the lambda runner have? Are you compiling in release mode?

256MB and yes, in the release mode.

> I was not able to recreate your example with either `--release` or debug compilation. I used your example, and modified to include some entities. (The default entity list from the cedarpolicy.com playground). Given the error message you've got, are you loading a lot of entities when you try it? (If it would help you/me debug, happy to chat more real-time on our slack)

I haven't managed to isolate it, so pointing to Axum example was the best bet, but the code's got more than that (custom extractors, state with AWS SDK clients etc.). No, it's not loading many (Cedar) entities (just `Entities::empty()` as in that sample code).
Is it this Slack: https://communityinviter.com/apps/cedar-policy/cedar-policy-language ? I applied but haven't got any invitation yet

**Comment by tomtau** [comment]
@aaronjeline we tried bumping up RAM on that lambda function to 4GB, but it didn't help.
I added logging of `stacker::remaining_stack()` before the Cedar policy authorizer call:
- remaining stack: Some(116240) -- went through
- remaining stack: Some(116496) -- went through
- remaining stack: Some(116240) -- went through
- remaining stack: Some(81504) -- failed / rejected with "error occurred while evaluating entity attributes: recursion limit reached" 

so it just happens that in that lambda environment, sometimes `stacker::remaining_stack()` would return a number below 100kb (the hardcoded limit enforced in Cedar) before calling the Cedar evaluation itself.

I found a workaround that I wrapped all those Cedar policy `is_authorized` calls in `stacker::grow`.
